### ISSUE TITLE
Packer 1.4.1 crash when using powershell provisioner across windows-openssh 7.9.0.1-Beta1

### ISSUE THREAD

**Issue by petemounce** [hashicorp/packer#7730]
When I attempt 

- Packer version from `packer version`
    
    ```
    $ packer version
    Packer v1.4.1
    ```

- Host platform `macOS Mojave 10.14.5`
- Debug log output from `PACKER_LOG=1 packer build template.json`: https://gist.github.com/petemounce/dd068b390c19c0d2df6d71c39c3b991c
- Sorry, I can't include the whole template. I've included the metadata part, which I think is the most likely. If you can ask for more information, I can assess internally whether I'm able to share other excerpts. I'm really sorry!

Target platform is Windows Server 2016 with OpenSSH 7.9.0.1-Beta1 (see gist for bootstrap script), using packer's ssh communicator.

**Comment by SwampDragons** [comment]
Looking at your logs, it looks like the failure is coming from a shell-local script calling Ansible, not a powershell script. Can you at least share that provisioner configuration with me?

From your logs (most of this is piped Ansible logs, not logs that originate within Packer): 
```
2019/06/10 16:31:25 ui: [0;32m    googlecompute: [34mremote:         Checking connectivity...[0m
2019/06/10 16:31:28 ui: [0;32m    googlecompute: [0m[1;31m104.199.90.181 | UNREACHABLE! => {[0m[0m
2019/06/10 16:31:28 ui: [0;32m    googlecompute: [1;31m    "changed": false, [0m[0m
2019/06/10 16:31:28 ui: [0;32m    googlecompute: [1;31m    "msg": "Data could not be sent to remote host \"104.199.90.181\". Make sure this host can be reached over ssh: Warning: Permanently added '104.199.90.181' (ECDSA) to the list of known hosts.\r\nno such identity: ./private_key: No such file or directory\r\npacker_user@104.199.90.181: Permission denied (publickey,password,keyboard-interactive).\r\n", [0m[0m
2019/06/10 16:31:28 ui: [0;32m    googlecompute: [1;31m    "unreachable": true[0m[0m
2019/06/10 16:31:28 ui: [0;32m    googlecompute: [1;31m}[0m[0m
2019/06/10 16:31:28 ui error: [1;31m==> googlecompute: panic: uncaught error[0m
2019/06/10 16:31:28 ui error: [1;31m==> googlecompute: Traceback (most recent call first):[0m
2019/06/10 16:31:28 ui error: [1;31m==> googlecompute:   at ansible/remote.sh:19 in check_connectivity()[0m
2019/06/10 16:31:28 ui error: [1;31m==> googlecompute:   at ansible/remote.sh:95 in main()[0m
2019/06/10 16:31:28 ui error: [1;31m==> googlecompute: ansible --module-name win_ping --inventory "../inventory/transient/${bake_version}.yml" "all" exited 4[0m
2019/06/10 16:31:28 [INFO] (telemetry) ending shell-local
2019/06/10 16:31:28 ui error: [1;31m==> googlecompute: Erroneous exit code 4 while executing script: /var/folders/w2/jd7cgbj92n9b3xbzkp5rft740000gn/T/packer-shell011429354

```

**Comment by petemounce** [comment]
@SwampDragons thanks for the fast response. It turns out that the ansible output is because in my inventory file, I had used `ansible_ssh_private_key_file` to specify a relative (and wrong) path to the temporary private key I was generating (I have not yet tried out to see whether the packer + googlecompute + ansible combination works together with windows+openssh, so I'm still using a shell-local to call ansible-playbook).

I've had this error several times - but never packer output tell me that there's a crash.log file and that the error was definitely a packer-side issue. The tail-end of the output:

```
==> googlecompute: Running local shell script: /var/folders/w2/jd7cgbj92n9b3xbzkp5rft740000gn/T/packer-shell011429354
    googlecompute: remote:         Installing python requirements for GCE dynamic inventory...
    googlecompute: Requirement already satisfied: wheel in /usr/local/lib/python3.7/site-packages (0.33.1)
    googlecompute: ~/src/ip/platform/packer/improbable.io/buildkite/ansible/buildkite-agent ~/src/ip/platform
    googlecompute: remote:         Checking connectivity...
    googlecompute: 104.199.90.181 | UNREACHABLE! => {
    googlecompute:     "changed": false,
    googlecompute:     "msg": "Data could not be sent to remote host \"104.199.90.181\". Make sure this host can be reached over ssh: Warning: Permanently added '104.199.90.181' (ECDSA) to the list of known hosts.\r\nno such identity: ./private_key: No such file or directory\r\npacker_user@104.199.90.181: Permission denied (publickey,password,keyboard-interactive).\r\n",
    googlecompute:     "unreachable": true
    googlecompute: }
==> googlecompute: panic: uncaught error
==> googlecompute: Traceback (most recent call first):
==> googlecompute:   at ansible/remote.sh:19 in check_connectivity()
==> googlecompute:   at ansible/remote.sh:95 in main()
==> googlecompute: ansible --module-name win_ping --inventory "../inventory/transient/${bake_version}.yml" "all" exited 4
==> googlecompute: Erroneous exit code 4 while executing script: /var/folders/w2/jd7cgbj92n9b3xbzkp5rft740000gn/T/packer-shell011429354
==> googlecompute:
==> googlecompute: Please see output above for more information.
==> googlecompute: Step "StepProvision" failed, aborting...
Build 'googlecompute' errored: unexpected EOF

==> Some builds didn't complete successfully and had errors:
--> googlecompute: unexpected EOF

==> Builds finished but no artifacts were created.
panic: uncaught error
2019/06/10 16:31:28 ui error: ==> googlecompute: Traceback (most recent call first):
2019/06/10 16:31:28 ui error: ==> googlecompute:   at ansible/remote.sh:19 in check_connectivity()
2019/06/10 16:31:28 ui error: ==> googlecompute:   at ansible/remote.sh:95 in main()
2019/06/10 16:31:28 ui error: ==> googlecompute: ansible --module-name win_ping --inventory "../inventory/transient/${bake_version}.yml" "all" exited 4
2019/06/10 16:31:28 [INFO] (telemetry) ending shell-local
2019/06/10 16:31:28 ui error: ==> googlecompute: Erroneous exit code 4 while executing script: /var/folders/w2/jd7cgbj92n9b3xbzkp5rft740000gn/T/packer-shell011429354
==> googlecompute:
==> googlecompute: Please see output above for more information.
2019/06/10 16:31:28 ui error: ==> googlecompute: Step "StepProvision" failed, aborting...
2019/06/10 16:31:28 [INFO] (telemetry) ending googlecompute
2019/06/10 16:31:28 ui error: Build 'googlecompute' errored: unexpected EOF
2019/06/10 16:31:28 machine readable: error-count []string{"1"}
2019/06/10 16:31:28 ui error:
==> Some builds didn't complete successfully and had errors:
2019/06/10 16:31:28 machine readable: googlecompute,error []string{"unexpected EOF"}
2019/06/10 16:31:28 ui error: --> googlecompute: unexpected EOF
2019/06/10 16:31:28 ui:
==> Builds finished but no artifacts were created.
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited
2019/06/10 16:31:28 [INFO] (telemetry) Finalizing.
2019/06/10 16:31:28 waiting for all plugin processes to complete...
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited
2019/06/10 16:31:28 /usr/local/bin/packer: plugin process exited



!!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Packer crashed! This is always indicative of a bug within Packer.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Packer[1] so that we can fix this.

[1]: https://github.com/hashicorp/packer/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

There's a `panic` in there.

**Comment by petemounce** [comment]
Oh! That's a red herring. That `panic: uncaught error` line is definitely at my end; we have a standard shell script include with a `trap` that does that. Does packer parse its own output for `panic` and treat that as a crash...?

**Comment by SwampDragons** [comment]
We don't; that PACKER CRASH is a definite indicator that we're not handling an exception somewhere that we should be. When I didn't see that line in your logs I assumed it was entirely on your end, not that you'd left it out. My bad. 

I'd still really benefit from your sharing the particular shell-local configuration that's causing this, with any sensitive variables just redacted.

**Comment by petemounce** [comment]
No worries. Here's the shell-local:

```json
    {
      "type": "shell-local",
      "command": "bash ansible/remote.sh buildkite-agent \"{{user `image_platform`}}\" \"{{user `version`}}\" {{user `ansible_verbosity`}} --skip-tags finalisations --extra-vars \"debug={{user `ansible_debug_output`}} agent_count={{user `agent_count`}} agent_is_root={{user `agent_is_root`}} agent_queue={{user `queue`}} bakery_binaries={{user `bakery_binaries`}} build_capability={{user `build_capability`}} image_platform={{user `image_platform`}} local_ssd_type={{user `local_ssd_type`}} permission_set={{user `permission_set`}}\""
    },
```

**Comment by SwampDragons** [comment]
Update: I'm still trying to figure out a repro case on my end. I don't necessarily need to see the entire remote.sh script, but I'm curious about how it's exiting after that trap and panic comment. 

My current vague theory is that your ansible script was doing something fun to the networking on the remote instance which caused the ssh communicator to disconnect in an unclean way.  But I'm a bit blind right now in terms of actually reproducing since I don't have a true repro case.

**Comment by petemounce** [comment]
We have a standard bash include; here are what I hope are the pertinent bits.

```bash
# configure cwd, vars and logging
_setup_env() {
  # -ET: propagate DEBUG/RETURN/ERR traps to functions and subshells
  # -e: exit on unhandled error
  # pipefail: any failure in a pipe causes the pipe to fail
  set -eET -o pipefail
  [[ -n "${DEBUG:-}" ]] && set -x
  trap _err_trap ERR

  if [[ "${#BASH_SOURCE[@]}" -ne 3 ]]; then
    echo -e "setup.sh:\tMust be sourced directly from an executed script."
    return 1
  fi

  readonly START_DIR="${PWD}"
  readonly SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[2]}")" && pwd)"
  readonly TOP_SCRIPT="${SCRIPT_DIR}/$(basename "${BASH_SOURCE[2]}")"
  if [[ -z "${SCRIPT_DIR}" ]]; then
    echo -e "setup.sh:\tFailed to determine directory containing executed script."
    return 1
  fi
  if ! cd "$(dirname "${BASH_SOURCE[0]}")/../.."; then
    echo -e "setup.sh:\tFailed to cd to repository root"
    return 1
  fi

  export GOPATH="${PWD}/go"
  export GOBIN="${GOBIN:-${GOPATH}/bin}"

  if ! source ci/lib/log.sh; then
    echo -e "setup.sh:\tFailed to source logging library"
    return 1
  fi
}

_err_trap() {
  local err=$?
  local cmd="${BASH_COMMAND:-}"
  # Disable echoing all commands as this makes the traceback really hard to follow
  set +x
  echo "panic: uncaught error" 1>&2
  print_traceback 1
  echo "${cmd} exited ${err}" 1>&2
}

# Print traceback of call stack, starting from the call location.
# An optional argument can specify how many additional stack frames to skip.
print_traceback() {
  local skip=${1:-0}
  local start=$((skip + 1))
  local end=${#BASH_SOURCE[@]}
  local curr=0
  echo "Traceback (most recent call first):" 1>&2
  for ((curr = start; curr < end; curr++)); do
    local prev=$((curr - 1))
    local func="${FUNCNAME[$curr]}"
    local file="${BASH_SOURCE[$curr]}"
    local line="${BASH_LINENO[$prev]}"
    echo "  at ${file}:${line} in ${func}()" 1>&2
  done
}
```

**Comment by SwampDragons** [comment]
So I went in and read our panic handler and I think this issue actually was just Packer seeing the line "panic:" and thinking it was a golang panic and that Packer itself had crashed. I'm going to close this since I _think_ that's what's going on and I was never able to get a repro case; if you want to not get this issue again you can change the "panic:" to literally any other string (even "panic -- " wouldn't have caused this).  

Sorry for the long delay on this!
